### ISSUE TITLE
nested structure for the slots

### ISSUE THREAD

**Issue by djarecka** [linkml/linkml#1846]
I'm wondering what is the best way to create a nested structure for slots, that would match `@nest` from `jsonld`. 

I know I can create an additional class, that would have slots from the nested list, but that would create an additional node in my graph.

I know that I can put `in_subset`, but this is not translated as `@nest` object when creating context or shacl.ttl files, so I'm getting errors when validating existing jsonld files that have this nested structure.

**Comment by cmungall** [comment]
Hi Dorota

I think what you are asking for is a kind of "pass-through" mechanism,
something like what is being requested here:

https://stackoverflow.com/questions/38876773/annotating-nested-structures-values-in-json-ld

(although I am not sure how in_subset might work)

Currently, the structure of the native LinkML object, the JSON/YAML tree,
and the RDF graph are all (by design) isomorphic. If there is an additional
structure in the JSON, then it has to be accounted for with a class that
models that intermediate structure.

Doing on-the-fly flattening is considered a separate concern, and is
intended to be handled by the (not yet production ready) linkml-transformer
framework.

There is an argument for allowing limited simple transformations in the
core framework itself. There is a precedent for this, in that mapping to
the relational model for SQL entails transformations, but these are limited
strictly to the necessary ones.

But how this would work for pass-throughs would need to be specced out; for
example, would the nesting be passed-through in the pydantic/dataclass
objects?

Things get complicated here. It turns out Pydantic V2 does have a feature
called alias paths that allows this:

https://docs.pydantic.dev/latest/concepts/fields/#aliaspath-and-aliaschoices

However, there isn't AFAIK a corresponding feature in dataclasses, or AFAIK
in the equivalent mechanisms in other programming languages. This means the
interface exposed would be fundamentally different, which could be
confusing. As a user I want to either *consistently* say "foo.meta.field=1"
or "foo.field=1" regardless of the context.

My general preference is to hide magic as far as possible, to keep
structures isomorphic, to keep interfaces consistent, and have the user do
the transform. However, I can see a case for an extension similar to
AliasPath. If we did this then I think the default behavior for generators
that map to a framework that doesn't have this construct should be to throw
an error, unless explicitly over-ridden. So it would be a "use at your own
risk" kind of feature.

TL;DR - for now, do any flattening or unflattening outside of
serialization/deserialization. Let's see what others think...

**Comment by djarecka** [comment]
I just thought I can link the specific context file and `@nest` structure I'm trying to represent in linkml. In [this context there is a nest structure `ui`](https://github.com/ReproNim/reproschema/blob/e0d83d05da24cb96493a529f597bbc7220b0cee3/contexts/generic#L110), and multiple other attributes, e.g. `order`, `message` are part of `ui`

**Comment by cmungall** [comment]
Closing this issue. LinkML maintains isomorphic data structures by design - nested intermediate structures would break this principle. For complex JSON-LD mappings that require `@nest` or similar constructs, linkml-transformer is the recommended approach.

— Claude (reviewed and approved by @cmungall)
